### PR TITLE
Mark IntList as deprecated; add C10_DEPRECATED_USING

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#16824 Mark IntList as deprecated; add C10_DEPRECATED_USING**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13978429/)

There was a big wooly yak getting the deprecated macros to work.
Gory details are in Deprecated.h

Differential Revision: [D13978429](https://our.internmc.facebook.com/intern/diff/D13978429/)

### CODE REVIEW

**Comment by mhubii** [comment]
What to use instead of the IntList? @ezyang

**Comment by ezyang** [comment]
IntArrayRef!

**Review by gchanan — APPROVED** [review]

File: c10/util/Deprecated.h [hunk 1]
```diff
@@ -1,16 +1,69 @@
...
+// error that it is ignored
+# define C10_DEPRECATED_USING [[deprecated]]
+#elif defined(__GNUC__)
+# define C10_DEPRECATED_USING __attribute__((deprecated))
 #else
-#warning "You need to implement C10_DEPRECATED for this compiler"
-#define C10_DEPRECATED(function) function
-#endif // defined(__GNUC__)
-#endif // defined(__cplusplus) && __cplusplus > 201402L
+# warning "You need to implement C10_DEPRECATED_USING for this compiler"
```
Comment (gchanan) [inline c10/util/Deprecated.h]: I didn't quite follow all the cases...are we actually using this?  Is it crazy to make this an error?

Comment (ezyang) [reply c10/util/Deprecated.h]: I wouldn't make it an error... everything will "work" in this case, the user will just not get deprecated warnings. Better not break the build for no good reason.

Comment (gchanan) [reply c10/util/Deprecated.h]: What about the first question?  Is this actually used in any of our builds?

Comment (ezyang) [reply c10/util/Deprecated.h]: No, this shouldn't be exercised by any of our builds.

**Review by smessmer — APPROVED** [review]
I'm fine with this if you want to land it, but also have an alternative proposal for you to consider:
Have no split between `C10_DEPRECATED` and `C10_DEPRECATED_USING`, but disable `C10_DEPRECATED` on nvcc. That disables it for a few more cases than your proposal, but makes it easier to use. Also, disabling this in nvcc probably doesn't matter too much because the same code is run through non-nvcc and will show warnings there anyhow.

**Comment by ezyang** [comment]
@smessmer Your suggestion is reasonable, but I don't want to turn off nvcc support entirely, since that means that we will not get deprecated warnings for any `cu` code, which can include host code.

**Pull request closed without merging**